File: JS 문제 풀이/프로그래머스 Lv.1/키패드 누르기.js
function solution(numbers, hand) {
    let answer = '';
    let l = [4,1];
    let r = [4,3];

    let arr = new Object();
    arr[0]=[4,2];
    arr[1]=[1,1],arr[2]=[1,2],arr[3]=[1,3];
    arr[4]=[2,1],arr[5]=[2,2],arr[6]=[2,3];
    arr[7]=[3,1],arr[8]=[3,2],arr[9]=[3,3];

    for(let num of numbers){

        if(num===1||num===4||num===7) {
            answer+='L';
            l=arr[num];
        }
        else if(num===3||num===6||num===9) {
            answer+='R';
            r=arr[num];
        }
        else{
            let dist_l = Math.abs(l[0]-arr[num][0])+Math.abs(l[1]-arr[num][1]);
            let dist_r = Math.abs(r[0]-arr[num][0])+Math.abs(r[1]-arr[num][1]);

            if(dist_l===dist_r){
                if(hand === 'right'){
                    answer+='R';
                    r=arr[num];
                }
                else{
                    answer+='L';
                    l=arr[num];
                }
            }
            else if(dist_l>dist_r){
                    answer+='R';
                    r=arr[num];

            }
            else {
                    answer+='L';
                    l=arr[num];

            }
        }
    }

    return answer;
}

File: JAVA 문제풀이/프로그래머스 Lv.1/키패드 누르기.java
import java.util.Arrays;
import java.util.Objects;

class Solution {
    public String solution(int[] numbers, String hand) {
        String answer = "";
        int leftr = 1,lefth = 3;
        int rightr = 3, righth = 3;

        for(int i=0;i<numbers.length;i++){

            if(numbers[i]==1||numbers[i]==4||numbers[i]==7){
                answer+='L';
                leftr = 1;
                lefth=numbers[i]/3;
            }
            else if(numbers[i]==3||numbers[i]==6||numbers[i]==9){
                answer+='R';
                rightr = 3;
                righth = (numbers[i]/3)-1;
            }
            else{
                if(numbers[i]==0) numbers[i]=9;
                int left_len = Math.abs(leftr-2)+Math.abs(lefth-numbers[i]/3);
                int right_len = Math.abs(rightr-2)+Math.abs(righth-numbers[i]/3);

                if(left_len>right_len){
                    answer+='R';
                    rightr=2;
                    righth=numbers[i]/3;
                }
                else if(left_len<right_len){
                    answer+='L';
                    leftr=2;
                    lefth=numbers[i]/3;
                }
                else{
                    if(Objects.equals(hand,"right")){
                        answer+='R';
                        rightr=2;
                        righth=numbers[i]/3;
                    }
                    else if(Objects.equals(hand,"left")){
                        answer+='L';
                        leftr=2;
                        lefth=numbers[i]/3;
                    }
                }
            }
        }
        return answer;
    }
}

// System.out.print
